<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Creating an extension package for i2dash</title>

<script src="i2dash-extension_files/header-attrs-2.7/header-attrs.js"></script>
<script src="i2dash-extension_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="i2dash-extension_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="i2dash-extension_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="i2dash-extension_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="i2dash-extension_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="i2dash-extension_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="i2dash-extension_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="i2dash-extension_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="i2dash-extension_files/navigation-1.1/tabsets.js"></script>
<link href="i2dash-extension_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="i2dash-extension_files/highlightjs-9.12.0/highlight.js"></script>
<link href="i2dash-extension_files/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="i2dash-extension_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div id="header">



<h1 class="title toc-ignore">Creating an extension package for i2dash</h1>
<h4 class="author">Jens Preussner</h4>
<a class="author_email" href="mailto:#"><a href="mailto:jens.preussner@mpi-bn.mpg.de" class="email">jens.preussner@mpi-bn.mpg.de</a></a>
</address>
<h4 class="author">Arsenij Ustjanzew</h4>
<a class="author_email" href="mailto:#"><a href="mailto:arsenij.ustjanzew@gmail.com" class="email">arsenij.ustjanzew@gmail.com</a></a>
</address>
<h4 class="date">28 März 2021</h4>

</div>


<p>Vignette last complied on 2021-03-28. i2dash is licenced under MIT + file LICENSE.</p>
<p><img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/i2dash_logo.png?raw=true" align="center" width="150px"/></p>
<div id="introduction-and-scope" class="section level1">
<h1>Introduction and scope</h1>
<p>i2dash allows to programmatically generate interactive and flexible web-based dashboards. At its heart, the <code>i2dashboard</code> class acts as the basic framework for managing metadata, input and pages. Currently i2dash supports plain text, images, R objects, such as HTML widgets, and functions as input. Such functions can be used to generate fully customized dashboard content, for example very specific and customizable visualizations. This allows data from specific disciplines to be visualized in a targeted manner and component functionality to be adapted to the needs of different users. Thus we used the principle to develop extensions shown below and created the package <a href="https://loosolab.github.io/i2dash.scrnaseq/">i2dash.scrnaseq</a>, which allows the visualization and exploration of single-cell RNA-seq data.</p>
<p>Besides flexibility, advantages of using functions are their reusability, the possibility to implement interactivity and the possibility to extend to entire pages.</p>
<p>In this vignette we will show how to extend the functionality of i2dash. We will develop a reusable component, that contains an interactive 3D scatter plot. In the ‘interactive’ mode of the dashboard the user can additionally select the data plotted on the axes. We chose to implement the extension in its own R package, which allows good code organization and easy sharing with others.</p>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/scatter3D_result_shiny.png?raw=true" alt="Figure 1: Final customized and reusable component with Shiny input widgets." width="80%" />
<p class="caption">
Figure 1: Final customized and reusable component with Shiny input widgets.
</p>
</div>
<div id="availability" class="section level2">
<h2>Availability</h2>
<p>The directories, files and code created during this vignette is also available for download from <a href="https://s3.mpi-bn.mpg.de/data-i2dash-2020/MyExtension.zip">our S3 storage</a>.</p>
</div>
<div id="directory-structure" class="section level2">
<h2>Directory structure</h2>
<p>For the new extension, we set up a directory structure of a source package (if you want to learn more about package structure, see <a href="https://r-pkgs.org/package-structure-state.html">here</a>).</p>
<p>In figure 2 you can see that the folder ‘MyExtension’ contains subfolders ‘R’, ‘inst’, and ‘man’. The ‘R’ folder contains the R scripts and functions. The Rmarkdown template of our component is stored under ‘MyExtension/inst/templates/’. Further the folder ‘MyExtension’ contains important metadata files, e.g. ‘DESCRIPTION’.</p>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/file_structure.png?raw=true" alt="Figure 2: Directory structure for an extension R package." width="80%" />
<p class="caption">
Figure 2: Directory structure for an extension R package.
</p>
</div>
<p>The easiest way to create the desired structure in RStudio is to click <strong>File</strong> -&gt; <strong>New Project</strong>. In the dialog we select <strong>New Directory</strong>, then <strong>R Package</strong>.</p>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/RStudio_new_project.png?raw=true" alt="Figure 3:" width="80%" />
<p class="caption">
Figure 3:
</p>
</div>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/RStudio_new_project2.png?raw=true" alt="Figure 4:" width="80%" />
<p class="caption">
Figure 4:
</p>
</div>
<p>To conclude, we set a package name, select the directory, where the new package structure will be created and click on <strong>Create Project</strong>.</p>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/RStudio_new_project3.png?raw=true" alt="Figure 5:" width="80%" />
<p class="caption">
Figure 5:
</p>
</div>
<p>RStudio creates the directory and several metadata files. To make this directory look like the one in figure 2, we delete the file ‘hello.R’ from the ‘R’ folder and create the subfolder ‘templates’ under ‘inst’.</p>
</div>
<div id="dependencies-reexport-of-magrittr-pipe-operators" class="section level2">
<h2>Dependencies &amp; reexport of magrittr pipe operators</h2>
<p>In the further course of this tutorial we need to install the following packages:</p>
<pre class="r"><code>install.packages(c(&quot;i2dash&quot;,&quot;magrittr&quot;,&quot;dplyr&quot;,&quot;assertive.types&quot;,&quot;plotly&quot;))</code></pre>
<p>Also, we need to create the file ‘reexports.R’ and reexport the pipe-operators of the magrittr package. In the ‘reexports.R’ file we write the following code:</p>
<pre class="r"><code>#&#39; magrittr forward-pipe operator
#&#39;
#&#39; See \code{\link[magrittr]{\%&gt;\%}}.
#&#39; @name %&gt;%
#&#39; @importFrom magrittr %&gt;%
#&#39; @export %&gt;%
NULL

#&#39; magrittr forward-backward-pipe operator
#&#39;
#&#39; See \code{\link[magrittr]{\%&lt;&gt;\%}}.
#&#39; @name %&lt;&gt;%
#&#39; @importFrom magrittr %&lt;&gt;%
#&#39; @export %&lt;&gt;%
NULL</code></pre>
</div>
<div id="concept-of-extensibility" class="section level2">
<h2>Concept of extensibility</h2>
<p>The ability of i2dash to generate R Markdown code can be extended with additional packages. Figure 6 demonstrates the concept of extensibility. In order to develop a component or a page with a pre-defined layout and specific composition of components as an extension, the developer needs to provide a method and an R Markdown template. The method takes the i2dashboard object and the data relevant for the component. It validates and stores the data in a suitable data structure in a file (e.g. RDS file). Furthermore the method generates an R Markdown string based on a template for the component or page and extends the i2dashboard object.</p>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/concept1.png?raw=true" alt="Figure 6: Concept of extending i2dash." width="80%" />
<p class="caption">
Figure 6: Concept of extending i2dash.
</p>
</div>
<p>Figure 7 shows this principle applied to our objectives. The method ‘scatter3D’ will be stored in the ‘scatter3d.R’ file. When the user applies this method, to add the component to his dashboard, the method will use the Rmarkdown template stored in the file ‘scatter3D.R’ to generate the component.</p>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/concept2.png?raw=true" alt="Figure 7: Our example project applied to this concept." width="80%" />
<p class="caption">
Figure 7: Our example project applied to this concept.
</p>
</div>
</div>
</div>
<div id="main-function" class="section level1">
<h1>Main function</h1>
<p>Now we create a new file ‘scatter3D.R’ and write the function that evaluates the input and saves it in an .Rds file for this component. The input for this function is:</p>
<ul>
<li><strong>i2dashboard</strong>: the i2dashboard class object to add the component.</li>
<li><strong>x</strong>, <strong>y</strong>, <strong>z</strong>, <strong>colour</strong>: This is the data used to plot in the graph. The user can provide nuemric vectors or data.frames. If data.frames are provided, the columns are the possible selection options in the Shiny dropdown menues.</li>
<li><strong>title</strong>, <strong>x_title</strong>, <strong>y_title</strong>, <strong>z_title</strong>: With these parameters the dashboard creator can further customize the component. These parapeters are optional and should be characters.</li>
</ul>
<pre class="r"><code>#&#39; Validate the data and create a Rmd string of a component containing a 3D scatter plot.
#&#39;
#&#39; @param i2dashboard The i2dashboard class object.
#&#39; @param x Numeric observations mapped to the x-axis (numeric vector or data.frame).
#&#39; @param y Numeric observations mapped to the y-axis (numeric vector or data.frame).
#&#39; @param z Numeric observations mapped to the z-axis (numeric vector or data.frame).
#&#39; @param x_title The title of the x-axis.
#&#39; @param y_title The title of the y-axis.
#&#39; @param z_title The title of the z-axis.
#&#39; @param colour The title of the y-axis.
#&#39; @param title The title of the component
#&#39;
#&#39; @return An Rmarkdown sting for the rendered component.
#&#39; @export
scatter3D &lt;- function(i2dashboard, x, y, z, colour = NULL, x_title = NULL, y_title = NULL, z_title = NULL, title = NULL){
  # see contetent below 
}
</code></pre>
<p>Inside the function, we need to validate the input:</p>
<pre class="r"><code>  library(assertive.types)
  library(dplyr)

  # validate required parameters

  # convert vectors to data.frames:
  if(is.numeric(x)) x &lt;- data.frame(X = x)
  if(is.numeric(y)) y &lt;- data.frame(Y = y)
  if(is.numeric(z)) z &lt;- data.frame(Z = z)
  if(is.character(colour) | is.numeric(colour) | is.factor(colour)) colour &lt;- data.frame(&quot;colour&quot; &lt;- colour)

  assertive.types::assert_is_data.frame(x)
  assertive.types::assert_is_data.frame(y)
  assertive.types::assert_is_data.frame(z)

  # select columns only containing numeric or integer values
  x %&lt;&gt;%
    as.data.frame() %&gt;%
    dplyr::select_if(function(col) is.integer(col) | is.numeric(col))
  y %&lt;&gt;%
    as.data.frame() %&gt;%
    dplyr::select_if(function(col) is.integer(col) | is.numeric(col))
  z %&lt;&gt;%
    as.data.frame() %&gt;%
    dplyr::select_if(function(col) is.integer(col) | is.numeric(col))

  # provide column names
  if(is.null(colnames(x))) colnames(x) &lt;- paste0(&quot;X_&quot;, 1:ncol(x))
  if(is.null(colnames(y))) colnames(y) &lt;- paste0(&quot;Y_&quot;, 1:ncol(y))
  if(is.null(colnames(z))) colnames(z) &lt;- paste0(&quot;Z_&quot;, 1:ncol(z))

  # check correct dimensions
  if(nrow(x) != nrow(y)) stop(&quot;The number of rows in &#39;x&#39; and &#39;y&#39; is not equal.&quot;)
  if(nrow(x) != nrow(z)) stop(&quot;The number of rows in &#39;x&#39; and &#39;z&#39; is not equal.&quot;)

  # validate optional parameters
  if(!is.null(colour)){
    assertive.types::assert_is_data.frame(colour)
    colour %&lt;&gt;%
      as.data.frame() %&gt;%
      dplyr::select_if(function(col) is.integer(col) | is.numeric(col) | is.factor(col))
    if(is.null(colnames(colour))) colnames(colour) &lt;- paste0(&quot;Color_&quot;, 1:ncol(colour))
    if(nrow(x) != nrow(colour)) stop(&quot;The number of rows in &#39;x&#39; and &#39;colour&#39; is not equal.&quot;)
  }
  # check titles
  if(!is.null(title)) assertive.types::is_character(title)
  if(!is.null(x_title)) assertive.types::is_character(x_title)
  if(!is.null(y_title)) assertive.types::is_character(y_title)
  if(!is.null(z_title)) assertive.types::is_character(z_title)</code></pre>
<p>Then the input is saved into an .Rds file. The <code>title</code> of the component as well as the <code>env_id</code> are not saved and will be provided as characters by knitting the template.</p>
<pre class="r"><code>  # Create random env id
  env_id &lt;- paste0(&quot;env_&quot;, stringi::stri_rand_strings(1, 6, pattern = &quot;[A-Za-z0-9]&quot;))

  # Create component environment and save as .rds file
  env &lt;- new.env()
  env$x &lt;- x
  env$y &lt;- y
  env$z &lt;- z
  env$colour &lt;- colour
  env$x_title &lt;- x_title
  env$y_title &lt;- y_title
  env$z_title &lt;- z_title

  saveRDS(env, file = file.path(i2dashboard@datadir, paste0(env_id, &quot;.rds&quot;)))</code></pre>
<p>The ‘scatter3D’ function saves the data in an environment file and uses the ‘scatter3D.Rmd’ template to render the R Markdown string i.e. replacing the placeholders in the template. In this case, the placeholder {{ title }} is replaced by the provided title string and {{ env_id }} is replaced by the random generated ID used for naming of the saved environment file.</p>
<p>At the end of the function we return the rendered R markdown string of the component.</p>
<pre class="r"><code>  # Generate the Rmarkdown string of the component based on the template
  timestamp &lt;- Sys.time()
  expanded_component &lt;- knitr::knit_expand(
    file = system.file(&quot;templates&quot;, &quot;scatter3D.Rmd&quot;, package = &quot;MyExtension&quot;),
    title = title,
    env_id = env_id,
    date = timestamp)
  return(expanded_component)</code></pre>
</div>
<div id="template-of-the-component" class="section level1">
<h1>Template of the component</h1>
<p>In the next step, we create an Rmarkdown template file of the scatter3D component under ‘inst/templates/scatter3D.Rmd’. At the top of this document the ‘title’ in curved brackets will be replaced by the provided title during knitting. Also the ‘env_id’ will be replaced and the correct .rds file will be read in. ‘is_shiny’ contains a logical whether the i2dashboard is interactive and should use shiny or not. Whether the dashboard is interactive or not is defined by the dashboard creator during the initialisation of the dashboard (the parameter <code>interactive</code> of the <code>i2dash::i2dashboard()</code> function). This is the first code chunk of the ‘scatter3D.Rmd’ file:</p>
<pre><code>
### {{ title }}

```{r}
{{ env_id }} <- readRDS(file.path(datadir, '{{ env_id }}.rds'))
is_shiny <- identical(knitr::opts_knit$get('rmarkdown.runtime'), 'shiny')
library(magrittr)
</code>
</pre>
<pre><code>
## The static mode

In this section, we will write the code chunk responsible for the static mode of the dashboard. The plotly plot will be interactive but there are no drop-down selections for selecting the data. The first column of each input data.frame (**x**, **y**, **z**, **colour**) will be used as the data for the 3D scatter plot.

&lt;div class=&quot;figure&quot;&gt;
&lt;img src=&quot;https://github.com/loosolab/i2dash/blob/master/vignettes/images/scatter3D_result_static.png?raw=true&quot; alt=&quot;Figure 8&quot; width=&quot;80%&quot; /&gt;
&lt;p class=&quot;caption&quot;&gt;Figure 8&lt;/p&gt;
&lt;/div&gt;

The following R code chunk is also in the same file `scatter3D.Rmd`. This code chunk will be executed, if Shiny is not used.

All variables we define in this code chunk should be unique. This is made possible by using the env_id in the curved brackets. During the knitting process, `{{ env_id }}` will be replaced by the &#39;real&#39; env_id and the variable names will be valid in the final R markdown string. Therefore, we can ignore RStudio warnings.

First we set the variables for `x`, `y`, `z`, `x_title`, `y_title`, and `z_title`. If the user had not provided his own titles, the names of the first column of `x`, `y` and `z` are used as the titles. If `colour` was provided the first column will be used as the input vector.

Next, the plotly chart is created by the `plotly::plot_ly()` function. Finally a data.frame is created, that will be then provided as download.


```{=html}
&lt;pre&gt;
```{r, eval=!is_shiny}
# set variables 
# the first column is always used
x_{{ env_id }} &lt;- {{ env_id }}$x[,1]
y_{{ env_id }} &lt;- {{ env_id }}$y[,1]
z_{{ env_id }} &lt;- {{ env_id }}$z[,1]
if(!is.null({{ env_id }}$x_title)) x_title_{{ env_id }} &lt;- {{ env_id }}$x_title else x_title_{{ env_id }} &lt;- colnames({{ env_id }}$x)[1]
if(!is.null({{ env_id }}$y_title)) y_title_{{ env_id }} &lt;- {{ env_id }}$y_title else y_title_{{ env_id }} &lt;- colnames({{ env_id }}$y)[1]
if(!is.null({{ env_id }}$z_title)) z_title_{{ env_id }} &lt;- {{ env_id }}$z_title else z_title_{{ env_id }} &lt;- colnames({{ env_id }}$z)[1]
if(!is.null({{ env_id }}$colour)) colour_{{ env_id }} &lt;- {{ env_id }}$colour[,1] else colour_{{ env_id }} &lt;- NULL

# creating the plot object
plot_{{ env_id }} &lt;- plotly::plot_ly(
  x = x_{{ env_id }}, 
  y = y_{{ env_id }}, 
  z = z_{{ env_id }}, 
  color = colour_{{ env_id }}) %&gt;%
  plotly::add_markers() %&gt;% 
  plotly::layout(scene = list(
    xaxis = list(title = x_title_{{ env_id }}),
    yaxis = list(title = y_title_{{ env_id }}),
    zaxis = list(title = z_title_{{ env_id }})
  ))

# Provide data for download
if(is.null(colour_{{ env_id }})){
  df_{{ env_id }} &lt;- data.frame(x = x_{{ env_id }}, 
                                y = y_{{ env_id }}, 
                                z = z_{{ env_id }})
} else {
  df_{{ env_id }} &lt;- data.frame(x = x_{{ env_id }}, 
                                y = y_{{ env_id }}, 
                                z = z_{{ env_id }},
                                colour = colour_{{ env_id }})
}
htmltools::div(style=&#39;display:block;float:left;width:100%;height:90%;&#39;,
  htmltools::tags$button(i2dash::embed_var(df_{{ env_id }})), 
  plot_{{ env_id }}
)</code></pre>
</pre>
<pre><code>
### Installing &amp; testing the static mode

Since we now have a lot of code, it makes sense to test if everything works so far. To do this we first have to generate the documentation of the package with the function `devtools::document()`. and then we install the package with `devtools::install()`. 


```r
devtools::document(&quot;.&quot;)
devtools::install(&quot;.&quot;, quick = TRUE)</code></pre>
<p>In a separate R file we create an i2dashboard object, add our component with <code>add_component()</code>, define the input and assemble the dashboard to a Rmarkdown file. If these steps have thrown no errors, half of our project has succeeded.</p>
<p>If we can open the newly generated file ‘MyDashboard.Rmd’ in RStudio and then view its contents in a web browser, this means that the template code also works. (Unfortunately, the 3D scatter plot of plotly is not displayed in the RStudio Viewer, so we need to view the file in the browser.)</p>
<pre class="r"><code>library(magrittr)
library(i2dash)

# Create i2dashboard object
# &amp; add the component to the default page
MyDashboard &lt;- i2dashboard(
  title= &quot;Test extension&quot;,
  datadir = &quot;.&quot;,
  interactive = FALSE
) %&gt;% add_component(
  component = MyExtension::scatter3D,
  x = mtcars$mpg,
  y = mtcars$wt,
  z = mtcars$hp,
  colour = as.factor(mtcars$cyl),
  title = &quot;My component&quot;)

# Assemble the dashboard
assemble(MyDashboard, file = &quot;MyDashboard.Rmd&quot;)</code></pre>
<div id="the-interactive-mode-with-r-shiny" class="section level2">
<h2>The interactive mode with R Shiny</h2>
<p>By adding Shiny to a flexdashboard, we create a dashboard that enables the user to change underlying parameters and see the results immediately. In our case we provide the possibility to select the values that should be mapped to the x-, y- and z-axis as well as the values, that should be used for the colouring.</p>
<p>A Shiny application consists of two parts, a user interface (UI) function, and a server function. The UI determines the structure and layout of the application and can contain HTML elements or input widgets for user-application interaction. Shiny also provides a group of functions that convert R objects into output widgets that are displayed on the UI. The server function defines the server-side logic of the Shiny application and is the “live” connection between input values on the client and outputs generated on the web server. The server function contains functions that map user inputs to various kinds of output.</p>
<p>In a flexdashboard the code for the UI as well as the server code can be used within the same code chunk.The webpage of flexdashboard provides a good description how to use Shiny with flexdashboard (see <a href="https://rmarkdown.rstudio.com/flexdashboard/shiny.html">here</a>).</p>
<div class="figure">
<img src="https://github.com/loosolab/i2dash/blob/master/vignettes/images/scatter3D_result_shiny.png?raw=true" alt="Figure 9: Final customized and reusable component with Shiny input widgets." width="80%" />
<p class="caption">
Figure 9: Final customized and reusable component with Shiny input widgets.
</p>
</div>
<p>First we create Shiny <code>selectInput()</code> widgets if the target data.frame has more than one column. Also a download button will be provided, so the user can download the data that is used for the created plot.</p>
<p>We handle the inputs by creating a reactive expression for each data.frame. Inside the reactives it is checked, if the data.frame has more than one column. If this is the case, the input values from the <code>selectInput()</code> widgets should be used.</p>
<p>Then a <code>downloadHandler()</code> creates a data.frame containing the currently selected x, y, z and colour values for the download.</p>
<p>Plotly’s function <code>renderPlotly()</code> outputs and renders a plotly object (generatet by the function <code>plotly::plot_ly()</code>) within Shiny. Finally, the UI elements are stored in a dropdown menu created with the package <a href="https://github.com/dreamRs/shinyWidgets">shinyWidgets</a>.</p>
<pre>
```{r, eval=is_shiny}
#
# shiny input widgets
#
ui_list <- list()

# shiny input widget for x
if (ncol({{ env_id }}$x) > 1){
  ui_list <- rlist::list.append(ui_list,
                                selectInput('input_x_{{ env_id }}', label = 'Select data for x axis:',
                                            choices = colnames({{ env_id }}$x)))
}

# shiny input widget for y
if (ncol({{ env_id }}$y) > 1){
  ui_list <- rlist::list.append(ui_list,
                                selectInput('input_y_{{ env_id }}', label = 'Select data for y axis:',
                                            choices = colnames({{ env_id }}$y)))
}

# shiny input widget for size
if (ncol({{ env_id }}$z) > 1){
  ui_list <- rlist::list.append(ui_list,
                                selectInput('input_z_{{ env_id }}', label = 'Select data for the z axis:',
                                            choices = colnames({{ env_id }}$z)))
}

# shiny input widget for colour_by
if (!is.null({{ env_id }}$colour)){
  if(ncol({{ env_id }}$colour) > 1)
  ui_list <- rlist::list.append(ui_list,
                                selectInput('input_colour_{{ env_id }}', label = 'Select metadata for colouring:',
                                            choices = colnames({{ env_id }}$colour)))
}

#
# shiny download button
#
ui_list <- rlist::list.append(ui_list, tags$div(tags$br(), downloadButton('downloadData_{{ env_id }}', 'Download data')))

#
# Handle inputs
#
x_{{ env_id }} <- shiny::reactive({
  if(ncol({{ env_id }}$x) == 1){
    data <- {{ env_id }}$x[[1]]
    title <- colnames({{ env_id }}$x)[1]
    return(list(data = data, title = title))
  } else {
    data <- {{ env_id }}$x[[input$input_x_{{ env_id }}]]
    title <- input$input_x_{{ env_id }}
    return(list(data = data, title = title))
  }
})

y_{{ env_id }} <- shiny::reactive({
  if(ncol({{ env_id }}$y) == 1){
    data <- {{ env_id }}$y[[1]]
    title <- colnames({{ env_id }}$y)[1]
    return(list(data = data, title = title))
  } else {
    data <- {{ env_id }}$y[[input$input_y_{{ env_id }}]]
    title <- input$input_y_{{ env_id }}
    return(list(data = data, title = title))
  }
})

z_{{ env_id }} <- shiny::reactive({
  if(ncol({{ env_id }}$z) == 1){
    data <- {{ env_id }}$z[[1]]
    title <- colnames({{ env_id }}$z)[1]
    return(list(data = data, title = title))
  } else {
    data <- {{ env_id }}$z[[input$input_z_{{ env_id }}]]
    title <- input$input_z_{{ env_id }}
    return(list(data = data, title = title))
  }
})

colour_{{ env_id }} <- shiny::reactive({
  if(!is.null({{ env_id }}$colour)){
    if(ncol({{ env_id }}$colour) == 1){
      return({{ env_id }}$colour[[1]])
    } else {
      return({{ env_id }}$colour[[input$input_colour_{{ env_id }}]])
    }
  } else {
    return(NULL)
  }
})

#
# Download data.frame
#
output$downloadData_{{ env_id }} <- downloadHandler(
  filename =  paste('data-', Sys.Date(), '.csv', sep=''),
  content = function(file) {
    if(is.null(colour_{{ env_id }}()$colour)){
      df <- data.frame(
        x = x_{{ env_id }}()$data, 
        y = y_{{ env_id }}()$data, 
        z = z_{{ env_id }}()$data)
    } else {
      df <- data.frame(
        x = x_{{ env_id }}()$data, 
        y = y_{{ env_id }}()$data, 
        z = z_{{ env_id }}()$data, 
        colour = colour_{{ env_id }}())
    }
    write.csv(df, file)
  }
)

#
# reactive for plot creation
#
output$plot_{{ env_id }} <- plotly::renderPlotly({
  if(!is.null({{ env_id }}$y_title)) y_title <- {{ env_id }}$y_title else y_title <- y_{{ env_id }}()$title
  if(!is.null({{ env_id }}$x_title)) x_title <- {{ env_id }}$x_title else x_title <- x_{{ env_id }}()$title
  if(!is.null({{ env_id }}$z_title)) z_title <- {{ env_id }}$z_title else z_title <- z_{{ env_id }}()$title

  plotly::plot_ly(
  x = x_{{ env_id }}()$data, 
  y = y_{{ env_id }}()$data, 
  z = z_{{ env_id }}()$data, 
  color = colour_{{ env_id }}()) %>%
  plotly::add_markers() %>% 
  plotly::layout(scene = list(
    xaxis = list(title = x_title),
    yaxis = list(title = y_title),
    zaxis = list(title = z_title)
  ))
})

#
# Layout of component
#
shiny::fillRow(flex = c(NA, 1),
      shinyWidgets::dropdownButton(div(style='max-height: 350px; overflow-x: auto;',do.call(shiny::inputPanel, ui_list)),
                       circle = TRUE, status = 'danger', icon = icon('gear'), width = '300px',
                       tooltip = shinyWidgets::tooltipOptions(title = 'Click, to change plot settings:'))
      ,
      plotly::plotlyOutput('plot_{{ env_id }}', height = '100%')
)
```</pre>
<div id="installing-testing-the-interactive-mode" class="section level3">
<h3>Installing &amp; testing the interactive mode</h3>
<p>To test the functionality we have to install the package with devtools as before.</p>
<pre class="r"><code>devtools::document(&quot;.&quot;)
devtools::install(&quot;.&quot;, quick = TRUE)</code></pre>
<p>Then we create a new i2dashboard object and add our component to it. This time we pass data.frames as input to the single parameters instead of single numeric vectors. Once the addittion and assemble steps are accomplished, we can open the Rmarkdown document and use RStudio’s “Run Document” button to render the file. In the web browser we should see the result, which is similar to the one in figure 1.</p>
<pre class="r"><code>library(magrittr)
library(i2dash)

# Create i2dashboard object
# &amp; add the component to the default page
MyDashboard &lt;- i2dashboard(
  title= &quot;Test extension&quot;,
  datadir = &quot;.&quot;,
  interactive = TRUE
) %&gt;% add_component(
  component = MyExtension::scatter3D,
  x = mtcars,
  y = mtcars,
  z = mtcars,
  colour = as.factor(mtcars$cyl),
  title = &quot;My component&quot;)

# Assemble the dashboard
assemble(MyDashboard, file = &quot;MyDashboard.Rmd&quot;)</code></pre>
</div>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="unnumbered">Session Info</h1>
<pre class="r"><code>sessionInfo()
#&gt; R version 4.0.4 (2021-02-15)
#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)
#&gt; Running under: Windows 10 x64 (build 19042)
#&gt; 
#&gt; Matrix products: default
#&gt; 
#&gt; locale:
#&gt; [1] LC_COLLATE=C                    LC_CTYPE=German_Germany.1252   
#&gt; [3] LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   
#&gt; [5] LC_TIME=German_Germany.1252    
#&gt; 
#&gt; attached base packages:
#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     
#&gt; 
#&gt; other attached packages:
#&gt; [1] magrittr_2.0.1   BiocStyle_2.18.1
#&gt; 
#&gt; loaded via a namespace (and not attached):
#&gt;  [1] digest_0.6.27       R6_2.5.0            jsonlite_1.7.2     
#&gt;  [4] evaluate_0.14       highr_0.8           rlang_0.4.10       
#&gt;  [7] stringi_1.5.3       jquerylib_0.1.3     bslib_0.2.4        
#&gt; [10] rmarkdown_2.7       tools_4.0.4         stringr_1.4.0      
#&gt; [13] xfun_0.20           yaml_2.2.1          compiler_4.0.4     
#&gt; [16] BiocManager_1.30.10 htmltools_0.5.1.1   knitr_1.31         
#&gt; [19] sass_0.3.1</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
